### PR TITLE
Handle shadowed plugins

### PR DESCRIPTION
**Pull request by brimoor**
## Change log

- Updates the plugins API to gracefully ignore plugins that are "shadowed" by other plugins of the same name
- Ensures that `download_plugin(..., overwrite=True)` and `create_plugin(..., overwrite=True)` cannot overwrite builtin plugins

## Example usage

```shell
PLUGINS_DIR="$(fiftyone config plugins_dir)"

# (if necessary) remove source install of `fiftyone-plugins`
unlink "$PLUGINS_DIR/fiftyone-plugins"

# Install a plugin
fiftyone plugins download \
    https://github.com/voxel51/fiftyone-plugins \
    --plugin-names @voxel51/dashboard

# Duplicate the plugin
ORIG_DIR="$PLUGINS_DIR/@voxel51/dashboard"
DUP_DIR="$PLUGINS_DIR/also_dashboard"
cp -R $ORIG_DIR $DUP_DIR

# See how duplicate plugins are handled
fiftyone plugins list
"""
plugin                            version    enabled    builtin    shadowed    directory
--------------------------------  ---------  ---------  ---------  ----------  ------------------------------------------------------
@voxel51/operators                1.0.0      ✓          ✓                      /Users/Brian/dev/fiftyone/plugins/operators
@voxel51/panels                   1.0.0      ✓          ✓                      /Users/Brian/dev/fiftyone/plugins/panels
@voxel51/dashboard                1.0.1      ✓                                 /Users/Brian/fiftyone/__plugins__/@voxel51/dashboard
@voxel51/dashboard                1.0.1                            ✓           /Users/Brian/fiftyone/__plugins__/also_dashboard
"""

# Verify that the App works
# EG launch the Dashboard panel
fiftyone app launch

# Only affects the "active" plugin
# "Shadowed" plugin remains disabled
fiftyone plugins disable @voxel51/dashboard
fiftyone plugins list

# Only affects the "active" plugin
# "Shadowed" plugin remains disabled
fiftyone plugins enable @voxel51/dashboard
fiftyone plugins list

# Shows the "active" plugin
# The `shadow_paths` row indicates that this plugin is shadowing `also_dashboard`
fiftyone plugins info @voxel51/dashboard

# Show info about a "shadowed" plugin
fiftyone plugins info $DUP_DIR

# Delete a shadowed plugin
rm -rf $DUP_DIR

# The `shadowed` column is only shown if there are actually shadowed plugins
fiftyone plugins list
"""
plugin                            version    enabled    builtin    directory
--------------------------------  ---------  ---------  ---------  ------------------------------------------------------
@voxel51/operators                1.0.0      ✓          ✓          /Users/Brian/dev/fiftyone/plugins/operators
@voxel51/panels                   1.0.0      ✓          ✓          /Users/Brian/dev/fiftyone/plugins/panels
@voxel51/dashboard                1.0.1      ✓                     /Users/Brian/fiftyone/__plugins__/@voxel51/dashboard
"""

# Cleanup
fiftyone plugins delete @voxel51/dashboard

# (if necessary)
cd /path/to/fiftyone-plugins
ln -s "$(pwd)" "$(fiftyone config plugins_dir)/fiftyone-plugins"
```

## Release Notes

- **New Features**
  - Enhanced plugin management with support for shadowed plugins.
  - Expanded `fiftyone plugins info` command to accept plugin name or directory path.

- **Documentation**
  - Updated CLI documentation to clarify plugin information retrieval.
  - Improved command usage examples for plugin-related commands.

- **Improvements**
  - Added more comprehensive plugin listing capabilities, including shadowed plugins.
  - Enhanced plugin detection and management logic.
  - Improved error handling and logging for plugin operations.

These updates provide more flexibility and clarity in managing FiftyOne plugins, making it easier to discover and interact with plugin resources.

### CODE REVIEW

**Comment by ritch** [comment]
```
(fiftyone-env2) ➜  plugins git:(prompt-on-cancel) ✗ fiftyone plugins disable @voxel51/panels
Traceback (most recent call last):
  File "/Users/ritchie/.pyenv/versions/fiftyone-env2/bin/fiftyone", line 8, in <module>
    sys.exit(main())
  File "/Users/ritchie/Projects/voxel51/fiftyone/fiftyone/core/cli.py", line 4715, in main
    args.execute(args)
  File "/Users/ritchie/Projects/voxel51/fiftyone/fiftyone/core/cli.py", line 4698, in <lambda>
    parser.set_defaults(execute=lambda args: command.execute(parser, args))
  File "/Users/ritchie/Projects/voxel51/fiftyone/fiftyone/core/cli.py", line 3986, in execute
    fop.disable_plugin(name)
  File "/Users/ritchie/Projects/voxel51/fiftyone/fiftyone/plugins/core.py", line 125, in disable_plugin
    raise ValueError(
ValueError: Cannot change enablement of builtin plugin '@voxel51/panels'
```

Lots of unrelated output, would have expected just the error message instead of the trace.

**Comment by ritch** [comment]
Manually tested duplicate plugins, duplicate builtins, disabling builtins (possibly unrelated to this PR!) ✅

**Comment by ritch** [comment]
Also tested duplicate builtins in the app ✅

File: fiftyone/plugins/core.py [hunk 1]
```diff
@@ -44,19 +44,22 @@ class PluginPackage:
...
+def list_plugins(enabled=True, builtin=False, shadowed=False):
+    """Lists available plugins.
 
     Args:
         enabled (True): whether to include only enabled plugins (True) or only
             disabled plugins (False) or all plugins ("all")
         builtin (False): whether to include only builtin plugins (True) or only
             non-builtin plugins (False) or all plugins ("all")
+        shadowed (False): whether to include only "shadowed" duplicate plugins
+            (True) or only usable plugins (False) or all plugins ("all")
```
Comment (ritch) [inline fiftyone/plugins/core.py:62]: Can `builtin` and `shadowed` be used together? The wording here is confusing.

File: fiftyone/plugins/core.py [hunk 2]
```diff
@@ -278,15 +296,21 @@ def download_plugin(url_or_gh_repo, plugin_names=None, overwrite=False):
...
-                etau.delete_dir(existing_dir)
-                plugin_dir = existing_dir
+                if existing_plugin.builtin:
+                    raise ValueError(
+                        f"Cannot overwrite builtin plugin '{plugin.name}'"
+                    )
+
+                logger.info(f"Overwriting existing plugin '{plugin.name}'")
+                plugin_dir = existing_plugin.directory
+                etau.delete_dir(plugin_dir)
```
Comment (ritch) [inline fiftyone/plugins/core.py:313]: When does this happen? I would hope that something like `fiftyone plugins download XXX --force` would be required to avoid accidentally deleting something. Possibly exists, just commenting here to make sure we double check this.

File: fiftyone/plugins/core.py [hunk 3]
```diff
@@ -621,24 +668,25 @@ def _iter_plugin_metadata_files(root_dir=None, strict=False):
...
-    plugin = None
-    for _plugin in _list_plugins(enabled=enabled, builtin=builtin):
-        if _plugin.name == name:
-            if check_for_duplicates and plugin is not None:
-                raise ValueError(f"Multiple plugins found with name '{name}'")
+def _get_plugin(name, enabled=None, builtin=None):
+    for plugin in _list_plugins(
+        enabled=enabled, builtin=builtin, shadowed=False
+    ):
+        if plugin.name == name:
```
Comment (ritch) [inline fiftyone/plugins/core.py:675]: We use this style all over and its not that efficient. Instead of loading everything and then searching for a name we should load only until we get the named plugin. Mostly concerned about unnecessary fs calls.

Maybe a `_find_plugin(name=, plugin_dir=)` would be useful for this.

Can handle such improvements elsewhere.

File: fiftyone/plugins/definitions.py [hunk 4]
```diff
@@ -194,6 +202,7 @@ def to_dict(self):
         """
         return {
             "name": self.name,
+            "builtin": self.builtin,
```
Comment (ritch) [inline fiftyone/plugins/definitions.py:205]: ?

Comment (ritch) [reply fiftyone/plugins/definitions.py:205]: I find that moving this arbitrarily adds to the OSS => teams conflicts.

**Review by ritch — APPROVED** [review]
LGTM

File: fiftyone/plugins/core.py [hunk 5]
```diff
@@ -44,19 +44,22 @@ class PluginPackage:
 
     name: str
     path: str
+    shadow_paths: Optional[list[str]] = None
```
[the remark this answers is not in the document]
Comment (brimoor) [reply fiftyone/plugins/core.py:47]: My official mantra is follow the existing convention in the module you're contributing to, so I had to do it 😄 

And... also had to do it because it's required when defining data classes 😉

File: fiftyone/plugins/core.py (hunk 2 above)
Comment (brimoor) [reply fiftyone/plugins/core.py:313]: Correct, user must specify `download_plugin(..., overwrite=True)` or `fiftyone plugins download ... --overwrite` in order to overwrite an existing plugin.

They'll get an error without forcing the overwrite.

File: fiftyone/plugins/core.py (hunk 3 above)
Comment (brimoor) [reply fiftyone/plugins/core.py:675]: yes refactoring `_list_plugins()` to be a generator would 2x average performance here.

Importantly, however, I believe all of the App's plugin usage goes through `build_plugin_contexts()`, which is cached via `@plugins_cache`.

File: fiftyone/plugins/definitions.py (hunk 4 above)
Comment (brimoor) [reply fiftyone/plugins/definitions.py:205]: I moved it so that `builtin` appears immediately below `name` when printing plugin info:

```
$ fiftyone plugins info @voxel51/operators
key                     value
----------------------  -----------------------------------------------------
name                    @voxel51/operators
builtin                 True
...
```

That's a bit hacky of course as it relies on dicts having a stable iteration order, which isn't necessarily guaranteed in general.

Comment (brimoor) [reply fiftyone/plugins/definitions.py:205]: +1 for not causing diffs that have no functional purpose though

Comment (ritch) [reply fiftyone/plugins/definitions.py:205]: Ah actually this makes sense, just didn't realize.
